torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+                                       " forward mode gradients. This is not allowed in strict mode.".format(i))
+                elif stage == "double_back":
+                    raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode.".format(i))
+                else:
+                    raise RuntimeError("The hessian of the user-provided function is independent of "
+                                       "entry {} in the grad_jacobian. This is not allowed in strict "
+                                       "mode as it prevents from using the double backward trick to "
+                                       "replace forward mode AD.".format(i))
+
+            grads_i = torch.zeros_like(refs[i])
+        else:
+            if strict and create_graph and not grads_i.requires_grad:
+                if "double" not in stage:
+                    raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+                else:
+                    raise RuntimeError("The hessian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+
+        res += (grads_i,)
+
+    return res
+
+# Public API
+
+def vjp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the output of ``func``. This argument is optional when
+            ``func``'s output contains a single element and (if it is not provided) will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict is ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
+            said inputs, which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
```

Comment:
are these named tuples?  I'd probably call these something like "func_output" and "vjp" or something.

Reply:
No they are not named tuples, just regular ones.